cmd/exo/state.go

```diff
@@ -0,0 +1,192 @@
+	Hidden: true,
+	Args:   cobra.MaximumNArgs(0),
+	RunE: func(cmd *cobra.Command, args []string) error {
+		return nil
+	},
+}
+
+var stateGetCmd = &cobra.Command{
+	Use:    "get <component>",
+	Short:  "Print component state",
+	Hidden: true,
+	Args:   cobra.ExactArgs(1),
+	RunE: func(cmd *cobra.Command, args []string) error {
+		componentRef := args[0]
+		ctx := newContext()
+		checkOrEnsureServer()
+		cl := newClient()
+
+		workspace := requireWorkspace(ctx, cl)
+		output, err := workspace.GetComponentState(ctx, &api.GetComponentStateInput{Ref: componentRef})
+		if err != nil {
+			return err
+		}
+
+		return jsonutil.PrettyPrintJSONString(os.Stdout, output.State)
+	},
+}
+
+var stateSetCmd = &cobra.Command{
+	Use:    "set <component>",
+	Short:  "Set component state",
+	Long:   "Set component state to the JSON received on stdin.",
+	Hidden: true,
+	Args:   cobra.ExactArgs(1),
+	RunE: func(cmd *cobra.Command, args []string) error {
+		componentRef := args[0]
+		ctx := newContext()
+		checkOrEnsureServer()
+		cl := newClient()
+
+		var newState map[string]interface{}
+		if err := json.NewDecoder(os.Stdin).Decode(&newState); err != nil {
+			return fmt.Errorf("reading state from stdin: %v", err)
+		}
+
+		workspace := requireWorkspace(ctx, cl)
+		_, err := workspace.SetComponentState(ctx, &api.SetComponentStateInput{
+			Ref:   componentRef,
+			State: jsonutil.MustMarshalString(newState),
+		})
+		return err
+	},
+}
+
+var stateClearCmd = &cobra.Command{
+	Use:    "clear <component>",
+	Short:  "Clear component state",
+	Hidden: true,
+	Args:   cobra.ExactArgs(1),
+	RunE: func(cmd *cobra.Command, args []string) error {
+		componentRef := args[0]
+		ctx := newContext()
+		checkOrEnsureServer()
+		cl := newClient()
+
+		workspace := requireWorkspace(ctx, cl)
+		_, err := workspace.SetComponentState(ctx, &api.SetComponentStateInput{
+			Ref:   componentRef,
+			State: "{}",
+		})
+		return err
+	},
+}
+
+var stateEditCmd = &cobra.Command{
+	Use:    "edit <component>",
+	Short:  "Edit component state",
+	Long:   "Edit component state using your preferred editor.",
+	Hidden: true,
+	Args:   cobra.ExactArgs(1),
+	RunE: func(cmd *cobra.Command, args []string) error {
+		componentRef := args[0]
+		ctx := newContext()
+		checkOrEnsureServer()
+		cl := newClient()
+
+		workspace := requireWorkspace(ctx, cl)
+		output, err := workspace.GetComponentState(ctx, &api.GetComponentStateInput{Ref: componentRef})
+		if err != nil {
+			return err
+		}
+
+		tmpfile, err := ioutil.TempFile("", "state.*.json")
+		if err != nil {
+			return fmt.Errorf("creating temporary file: %w", err)
+		}
+		defer os.Remove(tmpfile.Name())
+
+		if err := jsonutil.PrettyPrintJSONString(tmpfile, output.State); err != nil {
+			return fmt.Errorf("writing to temporary file: %w", err)
+		}
+		if err := tmpfile.Close(); err != nil {
+			return fmt.Errorf("closing temporary file: %w", err)
+		}
+
+		// TODO: Read from EDITOR, or guess from common editors.
```

Comment:
Is this not doing that?

Reply:
Oops - forgot to remove the TODO comment!

Reply:
I know that all three of us have done that about a half dozen times each over the last year :)